sable/utils/fs.rs

```diff
@@ -5,24 +5,36 @@ use std::path::{Path, PathBuf};
 use tokio::fs;
 use tokio::io::AsyncWriteExt;
 
-pub async fn atomic_write(path: impl AsRef<Path>, data: impl AsRef<[u8]>) -> Result<(), Error> {
-    let temp_path = PathBuf::from(shellexpand::full("~/.cache/sable/temp")?.into_owned());
-
-    if temp_path.try_exists().is_err() {
-        fs::create_dir_all(temp_path.parent().unwrap()).await?;
+pub async fn atomic_write(path: &Path, data: impl AsRef<[u8]>, mode: u32) -> Result<(), Error> {
+    let temp_path = PathBuf::from(shellexpand::full("~/.cache/sable/tmp")?.into_owned());
+    let file_path = temp_path.join(path);
+
+    // Make sure parent of the file exists
+    let parent = file_path.parent().unwrap();
+    if let Ok(false) | Err(_) = parent.try_exists() {
+        fs::create_dir_all(&parent).await?;
     }
 
     let mut file = fs::OpenOptions::new()
         .write(true)
         .create(true)
         .truncate(true)
-        .open(&temp_path)
+        .mode(mode)
+        .open(&file_path)
         .await?;
 
     file.write_all(data.as_ref()).await?;
     file.sync_all().await?;
 
-    fs::rename(&temp_path, path).await?;
+    match fs::rename(&file_path, path).await {
+        // If rename fails, it might be because its on a different mount point
+        // Try to copy it there instead
+        Err(_) => {
+            fs::copy(&file_path, path).await?;
+            fs::remove_file(&file_path).await?;
```

Comment:
least sus thing ever

Reply:
well, if you copy the file (not move it) you leave it behind, thats why remove it afterwards